### ISSUE TITLE
[explicit-module-boundary-types] Complains about inner function expressions in arrow functions

### ISSUE THREAD

**Issue by essenmitsosse** [typescript-eslint/typescript-eslint#2161]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": [2, {"allowExpressions": true}],
    "@typescript-eslint/explicit-module-boundary-types": [2]
  }
}
```
```TS
export const foo = (): (n: number) => string => (n) => String(n)
```

**Expected Result**

Shouldn't give an error, since this issue was fixed for regular functions (#1845).

**Actual Result**

Gives the same errors as for regular functions prior to 3.1.0
```
Missing return type on function. eslint(@typescript-eslint/explicit-module-boundary-types)
Argument 'n' should be typed. eslint(@typescript-eslint/explicit-module-boundary-types)
```

**Additional Info**

n/a

**Versions**

package | version
-- | --
@typescript-eslint/eslint-plugin | 3.1.0
@typescript-eslint/parser | 3.1.0
TypeScript | 3.9.3
ESLint | 7.1.0
node | 13.14.0
npm | 6.14.4
